server/src/main/java/org/elasticsearch/index/engine/SyncLoop.java

```diff
@@ -0,0 +1,215 @@
+        this.listeners = new PriorityQueue<>(1024, Comparator.comparing(Tuple::v1));
+        this.lastSyncedNanos = System.nanoTime();
+    }
+
+    public void maybeScheduleIncrementalFlush() {
+        if (translog.shouldIncrementalFlush()) {
+            if (promiseSemaphore.tryAcquire()) {
+                submitSyncTask(false);
+            }
+        }
+    }
+
+    public void scheduleSync(Translog.Location location, Consumer<Exception> listener) {
+        if (locationSynced(lastSyncedLocation, location)) {
+            listener.accept(null);
+            return;
+        }
+        synchronized (listeners) {
+            ++listenerCount;
+            listeners.add(new Tuple<>(location, preserveContext(listener)));
+        }
+        if (promiseSemaphore.tryAcquire()) {
+            long delay = Math.max(ONE_MILLIS - (System.nanoTime() - lastSyncedNanos), 0);
+            if (delay == 0) {
+                submitSyncTask(true);
+            } else {
+                threadPool.schedule(syncTask(true), TimeValue.timeValueNanos(delay), ThreadPool.Names.FLUSH);
+            }
+        }
+    }
+
+    private void submitSyncTask(boolean startWithSync) {
+        threadPool.executor(ThreadPool.Names.FLUSH).execute(syncTask(startWithSync));
+    }
+
+    private AbstractRunnable syncTask(boolean startWithSync) {
+        return new AbstractRunnable() {
+            @Override
+            public void onAfter() {
+                promiseSemaphore.release();
+                if (isListenersEmpty() == false && promiseSemaphore.tryAcquire()) {
+                    submitSyncTask(true);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn("Uncaught exception on the sync loop", e);
+            }
+
+            @Override
+            protected void doRun() {
+                runSyncLoop(startWithSync);
+            }
+        };
+    }
+
+    private void runSyncLoop(boolean startWithSync) {
+        assert promiseSemaphore.availablePermits() == 0;
+        lastSyncedLocation = translog.getLastSyncedLocation();
+        if (startWithSync) {
+            doSyncAndNotify();
+        } else {
+            try {
+                translog.flush();
+            } catch (Exception e) {
+                logger.debug("Exception when incrementally flushing translog", e);
+            }
+        }
+        while (isListenersEmpty() == false) {
+            doSyncAndNotify();
+        }
+    }
+
+    private void doSyncAndNotify() {
+        assert promiseSemaphore.availablePermits() == 0;
+        Exception exception = null;
+        ArrayList<Consumer<Exception>> syncedListeners = new ArrayList<>(listenerCount());
+        try {
+            drainListeners(syncedListeners, lastSyncedLocation);
+            completeListeners(syncedListeners, null);
+            syncedListeners.clear();
+            if (isListenersEmpty()) {
```

Comment:
If we add a listener while completing the already synced listeners, it looks like we always sync the translog, even if the listener added is also already synced? I wonder if this should be a boolean out of the `drainListeners` method?